### PR TITLE
test(engine): Test migration from Jest to Karma

### PR DESCRIPTION
**Pull request by pmdartus**
## Details

This PR removes all the unit tests in the LWC engine that directly depends on the DOM and moves them to Karma integration test suite. This work is needed as part of the engine decoupling from the DOM.

## Does this PR introduce breaking changes?

* ✅ `No, it does not introduce breaking changes.`

If yes, please describe the impact and migration path for existing applications.

## The PR fulfills these requirements:
* Have tests for the proposed changes been added? ✅ 
* Have you followed [these instructions](../CONTRIBUTING.md#-commit-message-conventions) to clearly describe the issue being fixed or feature enhanced? ✅

## GUS work item
W-7532195

### CODE REVIEW

File: packages/integration-karma/test/services/hooks/index.spec.js [hunk 1]
```diff
@@ -0,0 +1,72 @@
+import { createElement } from 'lwc';
```
Comment (ravijayaramappa) [inline packages/integration-karma/test/services/hooks/index.spec.js:1]:
```suggestion
import { createElement, register } from 'lwc';
```

File: packages/integration-karma/test/services/hooks/index.spec.js [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+import XHooks from 'x/hooks';
+
+const { register } = LWC;
```
Comment (ravijayaramappa) [inline packages/integration-karma/test/services/hooks/index.spec.js:5]: ~const { register } = LWC;~

Comment (ravijayaramappa) [reply packages/integration-karma/test/services/hooks/index.spec.js:5]: I may be missing some context here

File: packages/integration-karma/test/services/hooks/index.spec.js (hunk 1 above)
Comment (pmdartus) [reply packages/integration-karma/test/services/hooks/index.spec.js:1]: The compiler prevents from doing this. This is why I extract it from the IIFE.

File: packages/integration-karma/test/template/attribute-tab-index/index.spec.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+function validAttribute(value, expectedValue) {
+    it(`should render the tabindex attribute in the DOM when set to ${value}`, () => {
+        const elm = createElement('x-test', { is: Test });
+        elm.tabIndexValue = value;
+        document.body.appendChild(elm);
+
+        expect(elm.shadowRoot.querySelector('div').getAttribute('tabindex')).toBe(expectedValue);
+    });
+}
```
Comment (ravijayaramappa) [inline packages/integration-karma/test/template/attribute-tab-index/index.spec.js:13]:
```suggestion
    it.each(
        [0, '0'],
        [-1, '-1']
    )(`should render the tabindex attribute in the DOM when set to %d`, (value, expectedValue) => {
        const elm = createElement('x-test', { is: Test });
        elm.tabIndexValue = value;
        document.body.appendChild(elm);

        expect(elm.shadowRoot.querySelector('div').getAttribute('tabindex')).toBe(expectedValue);
    });
```

Comment (ravijayaramappa) [reply packages/integration-karma/test/template/attribute-tab-index/index.spec.js:13]: err, let me find the jasmine equivalent.

File: packages/integration-karma/test/rendering/slotting/index.spec.js [hunk 4]
```diff
@@ -0,0 +1,22 @@
+import { createElement } from 'lwc';
+
+import RenderCountParent from 'x/renderCountParent';
+
+// WIP: Need to figure out why this test doesn't pass.
```
Comment (ravijayaramappa) [inline packages/integration-karma/test/rendering/slotting/index.spec.js:5]: Add a TODO reference to https://github.com/salesforce/lwc/pull/1617 and proceed with this PR.

**Comment by pmdartus** [comment]
@ravijayaramappa I am done with the test migration, could you do another pass on the review?

File: packages/integration-karma/test/shadow-dom/EventTarget-methods/EventTarget.addEventListener.spec.js [hunk 5]
```diff
@@ -0,0 +1,11 @@
+it('should accept a listener config as second parameter', () => {
+    const handleEvent = jasmine.createSpy('handleEvent', () => {});
```
Comment (ravijayaramappa) [inline packages/integration-karma/test/shadow-dom/EventTarget-methods/EventTarget.addEventListener.spec.js:2]:
```suggestion
    const handleEvent = jasmine.createSpy();
```

File: packages/integration-karma/test/shadow-dom/event-in-shadow-tree/event-propagation.spec.js [hunk 6]
```diff
@@ -300,3 +300,69 @@ describe('event propagation in nested shadow tree', () => {
         ]);
     });
 });
+
+describe('Event.stopPropagation', () => {
+    it('should not invoke component handler is the propagation is stopped by a template handler', () => {
```
Comment (ravijayaramappa) [inline packages/integration-karma/test/shadow-dom/event-in-shadow-tree/event-propagation.spec.js:305]:
```suggestion
    it('should not invoke component handler if the propagation is stopped by a template handler', () => {
```

File: packages/integration-karma/test/shadow-dom/event-in-shadow-tree/event-propagation.spec.js [hunk 7]
```diff
@@ -300,3 +300,69 @@ describe('event propagation in nested shadow tree', () => {
...
+        ];
+        expect(logs).toEqual([
+            [nodes.span, nodes.span, composedPath],
+            [nodes.div, nodes.span, composedPath],
+        ]);
+    });
+});
+
+describe('Event.stopImmediatePropagation', () => {
+    it('should not invoke component handler is the propagation is stopped by a template handler', () => {
```
Comment (ravijayaramappa) [inline packages/integration-karma/test/shadow-dom/event-in-shadow-tree/event-propagation.spec.js:337]:
```suggestion
    it('should not invoke component handler if the propagation is stopped by a template handler', () => {
```

**Comment by pmdartus** [comment]
Thanks for the review @ravijayaramappa

**Pull request merged**